### ISSUE TITLE
Batched forward-mode JVP for parameter Jacobians sharing one primal sweep

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#577]
## Motivation

We want to train field surrogates on `(x, y, ∂y/∂x)` triples instead of `(x, y)` — gradient-augmented (Sobolev) training. For a map with **many outputs** (10³–10⁴ probe points × 3 components × frequencies) and **few inputs** (~10 geometry/material parameters), the full Jacobian costs `min(n_in, n_out)` linearised sweeps, so **forward mode with ~10 tangents is the right mode**; reverse mode is the wrong one here.

The economics of the whole approach hinge on one number: the cost of all 10 derivatives relative to one plain solve. In FDTD a JVP tangent propagation reuses the primal update coefficients, so a batched JVP over N tangents should share the expensive primal sweep and pay mainly in tangent-state memory and arithmetic — a few times one solve, not N times. Without a batched implementation, each tangent is a separate sweep and the approach has no advantage over simply running more simulations.

## Current state

No forward-mode support anywhere in the package — `grep -rn "jacfwd\|jax.jvp\|linearize" rfx/` returns nothing. `rfx/vmap_sweep.py` batches over *material values* (separate primal solves), which is a different thing: it does not produce derivatives.

## Proposed API

```python
value, jac = rfx.jacobian_fwd(
    sim_fn,              # params -> observable pytree (e.g. complex DFT field tensor)
    params,              # pytree of n_p scalars (geometry + material)
    tangents="identity", # or an explicit (n_t, n_p) tangent matrix
    batch_tangents=True, # share one primal sweep across all tangents
)
```

Returning `(value, jac)` in one call matters: the primal is needed anyway and recomputing it per tangent is the waste we are trying to avoid.

## Acceptance criteria

- [ ] `jacobian_fwd` returns a Jacobian matching central finite differences to a documented tolerance on at least one geometry parameter and one material parameter
- [ ] a reported cost profile: wall time and peak memory for `n_t` ∈ {1, 4, 10} versus one plain solve, on a 3-D example — this number is the deliverable, whatever it turns out to be
- [ ] works with the DFT field probes (`add_dft_plane_probe`) as the observable, not only scalar S-parameters
- [ ] documented interaction with the segmented rematerialisation path in `simulation.py`

## Notes

Reverse mode remains the right choice for *scalar* acquisition functionals in active learning, and that already works via the existing remat path. This request is specifically about the many-output Jacobian.

**Comment by bk-squared** [comment]
Priority note: this has dropped from blocking to nice-to-have on our side.

The direction moved from training a field surrogate on derivative-augmented data to differentiable design and analysis of the aperture coupling operator directly. That needs reverse mode on a *scalar* objective, which rfx already supports through the existing remat path, rather than a many-output Jacobian. See #579 for the specific bilinear-scalar objective that is now our blocker.

Batched forward-mode Jacobians remain genuinely useful — gradient-augmented surrogate training is still a defensible line if a fixed product family ever justifies the amortisation — so this is worth keeping open. It is just no longer on our critical path, and nobody should reprioritise other work for it on our account.

**Comment by bk-squared** [comment]
## GPU cost characterization — the column PR #631 promised

PR #631 shipped with CPU-only numbers and an explicit note that a GPU run would follow. Here it is. Measured with the merged `scripts/benchmark_jacobian_fwd.py` at commit 27c4fad on an RTX 4090 (VESSL runs 369367252824 and 369367252825, `nvcr.io/nvidia/jax:24.10-py3`), n_steps=120, n_p=10, float32, no point probes registered.

### n_t = 10 versus one plain solve

| | 33,635 cells | | 178,475 cells (5.3× cells) | |
|---|---|---|---|---|
| | **CPU** | **GPU** | **CPU** | **GPU** |
| wall time | 4.42× | 3.88× | 9.11× | 7.52× |
| flops (compiler estimate, per scan step) | 17.47× | 13.16× | 18.59× | 16.42× |
| XLA temp bytes (compiler estimate) | 10.35× | 10.06× | 10.40× | 10.84× |

### What this changes: not the conclusion

GPU is modestly better across the board, and it does **not** rescue the economics the issue hoped for. The `(1 + n_t)`-class cost holds on both backends, and the degradation with problem size reproduces: on GPU the n_t=10 wall ratio goes 3.88× → 7.52× for a 5.3× cell increase, tracking the same drift toward the flop ratio that the CPU measurement showed. Marginal cost per tangent, in units of one plain solve: GPU 0.15 (small) → 0.61 (large); CPU 0.33 → 0.76.

**The small-grid GPU number is flattering and should not be quoted on its own.** A 3.88× reading at 33,635 cells does not survive a 5.3× grid increase — which is exactly why the repo requires an invariance witness before a measured number is published, and why every ratio above carries its grid size.

Batching still earns its keep, but by less at scale: batched-versus-sequential at n_t=10 is 5.4× on the small GPU grid and 2.26× on the large one (CPU: 3.1×–3.6×).

### What is backend-independent

The memory argument, which is the real case for forward mode here, does not depend on the backend: the n_t=10 allocation ratio is 10.06–10.84× across both backends and both grids, and it stays independent of `n_steps` (65,763,716 → 65,766,148 bytes from 120 to 240 steps, 0.004%). Reverse mode without segmented remat was 54.5× one solve at only 120 steps, and `jax.jacrev` cannot produce this Jacobian at all — it refuses the complex DFT tensor outright.

### One reading caveat on the script's own witness

The benchmark prints an `intercept/plain` ratio from a two-point fit as a primal-sharing witness and expects it near 1.0. On CPU it reads 0.98–1.10; on GPU it reads 1.41–1.87. That is **not** evidence that sharing broke. The fit pushes any fixed per-call overhead that does not scale with `n_t` — kernel launch, tiling and padding — into the intercept, and the effect is largest precisely where the marginal cost is smallest, which is the GPU. The reliable witness is the structural one and it is backend-independent: the `vmap(jvp)` jaxpr contains exactly one scan whose carry holds a single unbatched copy of the field state at every `n_t`, with the tangent carry scaling exactly ×n_t (gated by G3 in `tests/test_jacobian_fwd.py`). Worth tightening the script's expectation text so a future reader does not misread a GPU intercept as a regression.

Raw logs: `.omx/jacfwd-bench/20260810T145141Z-27c4fad/` and `.omx/jacfwd-bench-large/…` in the primary checkout.
